llcppg: intmax_t... type not exist;nil underlying type panic

```c
#  include <stdint.h>
typedef intmax_t ossl_intmax_t;
typedef uintmax_t ossl_uintmax_t;
# endif
```

2024/12/13 18:09:59 NewTypedefDecl ossl_intmax_t Fail: intmax_t not found sys type intmax_t in _types/_intmax_t.h not found in package github.com/goplus/llgo/c, full path /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/_types/_intmax_t.h
2024/12/13 18:09:59 NewTypedefDecl ossl_uintmax_t Fail: uintmax_t not found sys type uintmax_t in _types/_uintmax_t.h not found in package github.com/goplus/llgo/c, full path /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/_types/_uintmax_t.h
panic: underlying type must not be nil

--- a/cmd/gogensig/convert/_testdata/funcrefer/conf/llcppg.symb.json
+++ b/cmd/gogensig/convert/_testdata/funcrefer/conf/llcppg.symb.json
@@ -13,5 +13,10 @@
     "mangle": "OSSL_provider_init",
     "c++": "OSSL_provider_init",
     "go": "ProviderInit"
+  },
+  {
+    "mangle": "OSSL_PROVIDER_add_builtin",
+    "c++": "OSSL_PROVIDER_add_builtin",
+    "go": "ProviderAddBuiltin"
   }
 ]

--- a/cmd/gogensig/convert/_testdata/funcrefer/gogensig.expect
+++ b/cmd/gogensig/convert/_testdata/funcrefer/gogensig.expect
@@ -36,11 +36,20 @@ type OSSLProviderInitFn2 func(*OSSLCOREHANDLE, *OSSLDISPATCH, **OSSLDISPATCH, *u
 //go:linkname ProviderInit C.OSSL_provider_init
 func ProviderInit(*OSSLCOREHANDLE, *OSSLDISPATCH, **OSSLDISPATCH, *unsafe.Pointer) c.Int
 
+type OsslLibCtxSt struct {
+	Unused [8]uint8
+}
+type OSSLLIBCTX OsslLibCtxSt
+//go:linkname ProviderAddBuiltin C.OSSL_PROVIDER_add_builtin
+func ProviderAddBuiltin(*OSSLLIBCTX, name *int8, init_fn *OSSLProviderInitFn2) c.Int
+
 ===== llcppg.pub =====
 CallBack
 Hooks
 OSSL_CORE_HANDLE OSSLCOREHANDLE
 OSSL_DISPATCH OSSLDISPATCH
+OSSL_LIB_CTX OSSLLIBCTX
 OSSL_provider_init_fn OSSLProviderInitFn
 OSSL_provider_init_fn2 OSSLProviderInitFn2
 Stream
+ossl_lib_ctx_st OsslLibCtxSt

--- a/cmd/gogensig/convert/_testdata/funcrefer/hfile/temp.h
+++ b/cmd/gogensig/convert/_testdata/funcrefer/hfile/temp.h
@@ -26,3 +26,9 @@ typedef int(OSSL_provider_init_fn2)(const OSSL_CORE_HANDLE *handle,
                                    void **provctx);
 
 OSSL_provider_init_fn2 OSSL_provider_init;
+
+typedef struct ossl_lib_ctx_st OSSL_LIB_CTX;
+
+int OSSL_PROVIDER_add_builtin(OSSL_LIB_CTX *, const char *name,
+                              OSSL_provider_init_fn2 *init_fn);
+

--- a/cmd/gogensig/convert/convert_test.go
+++ b/cmd/gogensig/convert/convert_test.go
@@ -386,7 +386,6 @@ import (
 )
 
 type NormalType c.Int
-
 type Foo struct {
 	Unused [8]uint8
 }

--- a/cmd/gogensig/convert/package.go
+++ b/cmd/gogensig/convert/package.go
@@ -39,6 +39,7 @@ type Package struct {
 	cvt        *TypeConv                  // package type convert
 	curFile    *HeaderFile                // current processing c header file.
 	incomplete map[string]*gogen.TypeDecl // originname -> TypeDecl
+	deferTypes map[*gogen.TypeDecl]func() (types.Type, error)
 }
 
 type PackageConfig struct {
@@ -56,6 +57,7 @@ func NewPackage(config *PackageConfig) *Package {
 		p:          gogen.NewPackage(config.PkgPath, config.Name, config.GenConf),
 		conf:       config,
 		incomplete: make(map[string]*gogen.TypeDecl),
+		deferTypes: make(map[*gogen.TypeDecl]func() (types.Type, error)),
 	}
 
 	mod, err := gopmod.Load(config.OutputDir)
@@ -312,19 +314,55 @@ func (p *Package) NewTypedefDecl(typedefDecl *ast.TypedefDecl) error {
 	p.CollectNameMapping(typedefDecl.Name.Name, name)
 
 	genDecl := p.p.NewTypeDefs()
+	typeSpecdecl := genDecl.NewType(name)
+
+	if changed {
+		substObj(p.p.Types, p.p.Types.Scope(), typedefDecl.Name.Name, typeSpecdecl.Type().Obj())
+	}
+
+	if tagRef, ok := typedefDecl.Type.(*ast.TagExpr); ok {
+		inc := p.handleTyperefIncomplete(tagRef, typeSpecdecl)
+		if inc {
+			return nil
+		}
+	}
+
 	typ, err := p.ToType(typedefDecl.Type)
 	if err != nil {
+		typeSpecdecl.InitType(p.p, types.NewStruct(p.cvt.defaultRecordField(), nil))
 		return err
 	}
-	typeSpecdecl := genDecl.NewType(name)
+
 	typeSpecdecl.InitType(p.p, typ)
 	if _, ok := typ.(*types.Signature); ok {
 		genDecl.SetComments(NewTypecDocComments())
 	}
-	if changed {
-		substObj(p.p.Types, p.p.Types.Scope(), typedefDecl.Name.Name, typeSpecdecl.Type().Obj())
-	}
+
 	return nil
+}
+
+func (p *Package) handleTyperefIncomplete(tagRef *ast.TagExpr, typeSpecdecl *gogen.TypeDecl) bool {
+	var name string
+	switch n := tagRef.Name.(type) {
+	case *ast.Ident:
+		name = n.Name
+	case *ast.ScopingExpr:
+		// todo(zzy):scoping
+		panic("todo:scoping expr not supported")
+	}
+	_, inc := p.incomplete[name]
+	if !inc {
+		return false
+	}
+	p.deferTypes[typeSpecdecl] = func() (types.Type, error) {
+		typ, err := p.ToType(tagRef)
+		if err != nil {
+			return nil, err
+		}
+		// a function type will not be a incomplete type,so we not need to check signature and add comments
+		return typ, nil
+	}
+	return true
 }
 
 // Convert ast.Expr to types.Type
@@ -480,7 +518,17 @@ func (p *Package) WriteToBuffer(genFName string) (*bytes.Buffer, error) {
 	for _, decl := range p.incomplete {
 		decl.InitType(p.p, types.NewStruct(p.cvt.defaultRecordField(), nil))
 	}
+	for decl, getTyp := range p.deferTypes {
+		typ, err := getTyp()
+		if typ != nil {
+			decl.InitType(p.p, typ)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
 	p.incomplete = make(map[string]*gogen.TypeDecl, 0)
+	p.deferTypes = make(map[*gogen.TypeDecl]func() (types.Type, error), 0)
 	buf := new(bytes.Buffer)
 	err := p.p.WriteTo(buf, genFName)
 	if err != nil {

--- a/cmd/gogensig/convert/package_bulitin_test.go
+++ b/cmd/gogensig/convert/package_bulitin_test.go
@@ -1,0 +1,56 @@
+package convert
+
+import (
+	"testing"
+
+	"github.com/goplus/gogen"
+	"github.com/goplus/llcppg/ast"
+	cfg "github.com/goplus/llcppg/cmd/gogensig/config"
+	cppgtypes "github.com/goplus/llcppg/types"
+)
+
+func TestTypeRefIncompleteFail(t *testing.T) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Fatal("Expected panic, got nil")
+		}
+	}()
+	pkg := NewPackage(&PackageConfig{
+		PkgBase: PkgBase{
+			PkgPath:  ".",
+			CppgConf: &cppgtypes.Config{},
+			Pubs:     make(map[string]string),
+		},
+		Name:        "testpkg",
+		GenConf:     &gogen.Config{},
+		OutputDir:   "",
+		SymbolTable: cfg.CreateSymbolTable([]cfg.SymbolEntry{}),
+	})
+	pkg.cvt.SysTypeLoc["Bar"] = &HeaderInfo{
+		IncPath: "Bar",
+		Path:    "Bar",
+	}
+	pkg.incomplete["Bar"] = &gogen.TypeDecl{}
+	err := pkg.NewTypedefDecl(&ast.TypedefDecl{
+		Name: &ast.Ident{Name: "Foo"},
+		Type: &ast.TagExpr{
+			Name: &ast.Ident{Name: "Bar"},
+		},
+	})
+	if err != nil {
+		t.Fatal("NewTypedefDecl failed:", err)
+	}
+	delete(pkg.incomplete, "Bar")
+
+	_, err = pkg.WriteToBuffer("testpkg")
+	if err == nil {
+		t.Fatal("Expected error, got nil")
+	}
+
+	pkg.handleTyperefIncomplete(&ast.TagExpr{
+		Tag: 0,
+		Name: &ast.ScopingExpr{
+			X: &ast.Ident{Name: "Bar"},
+		},
+	}, nil)
+}

